W-14153679-idempotent-message-validator (#184)
mulesoft/docs-code-builder: modules/ROOT/pages/_partials/mule-components/4.4/component-idempotent-message-validator.adoc

//
//tag::component-title[]

= Idempotent Message Validator (<idempotent-message-validator/>)

//end::component-title[]
//

//
//tag::component-short-description[]
//     Short description of the form "Do something..." 
//     Example: "Configure log messages anywhere in a flow."

> Ensure that only unique messages continue through a flow execution by checking the unique ID of the incoming message.

Reviewer: check this description; component just seems to be establishing a unique identifier for a message but filtering could be one use case

> Ensures that only unique messages continue through a flow execution by checking the unique ID of the incoming message.

Reviewer: ```suggestion
Ensure that only unique messages continue through a flow execution by checking the unique ID of the incoming message.
```

|===
//end::component-xml-attributes-root[]
//
//
//TABLE (IF NEEDED): CHILD XML ATTRIBUTES for each child element
//  Repeat as needed, adding the next number to the tag value. 
//  Provide intro text, as needed.
//tag::component-xml-child1[]

[%header, cols="1,3"]
|===
| <CHILD_ELEMENT/> Attributes | Description

| `attribute_here` | DESCRIPTION_HERE_WITH_DEFAULT_IF_AVAILABLE Defaults to `DEFAULT_HERE`.
| `attribute_here` | DESCRIPTION_HERE_WITH_DEFAULT_IF_AVAILABLE Defaults to `DEFAULT_HERE`.

|===
//end::component-xml-child1[]
//
//
//TABLE (IF NEEDED): GRANDCHILD XML ATTRIBUTES for each grandchild element
//  Repeat as needed, adding the next number to the tag value. 
//  Provide intro text, as needed.
//TAG
//tag::component-xml-descendant1[]
[%header, cols="1,3"]
|===
| <GRANDCHILD_ELEMENT/> Attributes | Description

| `attribute_here` | DESCRIPTION_HERE_WITH_DEFAULT_IF_AVAILABLE Defaults to `DEFAULT_HERE`.
| `attribute_here` | DESCRIPTION_HERE_WITH_DEFAULT_IF_AVAILABLE Defaults to `DEFAULT_HERE`.

|===
//end::component-xml-descendant1[]
//

//SECTION: EXAMPLES
//
//tag::component-examples-title[]

== Examples

> The following examples show how to configure an Idempotent Message Validator to extract an ID from an HTTP request to use as a the unique identifier and how to hash the payload using a DataWeave expression and use the message ID as the unique identifier. 

Reviewer: ```suggestion
The examples show how to configure an Idempotent Message Validator to extract an ID from an HTTP request for use as a unique identifier and how to hash the payload using a DataWeave expression and use the message ID as the unique identifier. 
```

//end::component-examples-title[]
//
//
//tag::component-xml-ex1[]
[[example1]]

=== Example: Generate an ID from an HTTP Request

> The following example shows an Idempotent Message Validator configured to extract the query parameter id from the HTTP request and use that value as a unique identifier to filter the message:

Reviewer: ```suggestion
The following example shows an Idempotent Message Validator configured to extract the query parameter ID from the HTTP request and use that value as a unique identifier to filter the message:
```

[source,xml]
----
<flow name="myFlow">
  <http:listener doc:name="Listener" config-ref="HTTP_Listener_config" path="/"/>
  ...
  <idempotent-message-validator doc:name="Idempotent Message Validator" idExpression="#[attributes.queryParams.id]"/>
  ...
</flow>
----

//OPTIONAL: SHOW OUTPUT IF HELPFUL
//The example produces the following output: 

//OUTPUT_HERE 

//end::component-xml-ex1[]
//
//
//tag::component-xml-ex2[]
[[example2]]

=== Example: Generate an ID by Hashing the Payload With DataWeave

> The following example shows an Idempotent Message Validator configured to hash the payload using a DataWeave expression, store the result in a persistent object store, and then use the processed message IDs as unique identifiers for filtering the message:

Reviewer: don't see the filtering part

> The following example shows an Idempotent Message Validator configured to hash the payload using a DataWeave expression, store the result in a persistent object store, and then use the processed message IDs as the unique identifier to filter the message:

Reviewer: ```suggestion
The following example shows an Idempotent Message Validator configured to hash the payload using a DataWeave expression, store the result in a persistent object store, and then use the processed message IDs as unique identifiers for filtering the message:
```

[source,xml]
----
<flow name="myFlow">
  <http:listener doc:name="Listener" config-ref="HTTP_Listener_config" path="/"/>
  ...
  <idempotent-message-validator doc:name="Idempotent Message Validator" idExpression="
    #[%dw 2.0
      import dw::Crypto
      output application/octet-stream
      ---

>       Crypto::hashWith(payload,'MD5')]">

Reviewer: check this example; see the location of the closing square bracket
